homeassistant/components/upcloud/__init__.py

```diff
@@ -52,65 +64,200 @@
-        _LOGGER.debug("Updating UpCloud component")
-        hass.data[DATA_UPCLOUD].update()
-        dispatcher_send(hass, SIGNAL_UPDATE_UPCLOUD)
+    except requests.exceptions.RequestException as err:
+        _LOGGER.error("Failed to connect", exc_info=True)
+        raise ConfigEntryNotReady from err
+
+    upcloud_data = hass.data.setdefault(DATA_UPCLOUD, UpCloudHassData())
+
+    # Handle pre config entry (0.116) scan interval migration to options
+    migrated_scan_interval = upcloud_data.scan_interval_migrations.pop(
+        config_entry.data[CONF_USERNAME], None
+    )
+    if migrated_scan_interval and (
+        not config_entry.options.get(CONF_SCAN_INTERVAL)
+        or config_entry.options[CONF_SCAN_INTERVAL] == DEFAULT_SCAN_INTERVAL.seconds
+    ):
+        update_interval = migrated_scan_interval
+        hass.config_entries.async_update_entry(
+            config_entry,
+            options={CONF_SCAN_INTERVAL: update_interval.seconds},
+        )
+    elif config_entry.options.get(CONF_SCAN_INTERVAL):
+        update_interval = timedelta(seconds=config_entry.options[CONF_SCAN_INTERVAL])
+    else:
+        update_interval = DEFAULT_SCAN_INTERVAL
+
+    coordinator = upcloud_data.coordinators[
+        config_entry.data[CONF_USERNAME]
+    ] = UpCloudDataUpdateCoordinator(
+        hass,
+        update_interval=update_interval,
+        cloud_manager=manager,
+        username=config_entry.data[CONF_USERNAME],
+    )
+
+    # Listen to config entry updates
+    coordinator.unsub_handlers.append(
+        config_entry.add_update_listener(_async_signal_options_update)
+    )
+    coordinator.unsub_handlers.append(
+        async_dispatcher_connect(
+            hass,
+            _config_entry_update_signal_name(config_entry),
+            coordinator.async_update_config,
+        )
+    )
 
     # Call the UpCloud API to refresh data
-    upcloud_update(dt.utcnow())
-    track_time_interval(hass, upcloud_update, scan_interval)
+    await coordinator.async_request_refresh()
+    if not coordinator.last_update_success:
+        raise ConfigEntryNotReady
+
+    # Forward entry setup
+    for domain in CONFIG_ENTRY_DOMAINS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, domain)
+        )
 
     return True
 
 
-class UpCloud:
-    """Handle all communication with the UpCloud API."""
+async def async_unload_entry(hass, config_entry):
+    """Unload the config entry."""
+    for domain in CONFIG_ENTRY_DOMAINS:
+        await hass.config_entries.async_forward_entry_unload(config_entry, domain)
 
-    def __init__(self, manager):
-        """Initialize the UpCloud connection."""
-        self.data = {}
-        self.manager = manager
+    coordinator: UpCloudDataUpdateCoordinator = hass.data[
+        DATA_UPCLOUD
+    ].coordinators.pop(config_entry.data[CONF_USERNAME])
+    while coordinator.unsub_handlers:
+        coordinator.unsub_handlers.pop(0)()
```

Comment:
`pop(0)` is expensive. Better just use pop. Order doesn't matter. If you want to do it on order, just use `for unsub in unsub_handlers: unsub()`

Reply:
Actually in retrospect, in principle removing in reverse order compared to appending sounds better, and plain `pop()` satisfies that too, will do.